File: server/src/main/com/broll/gainea/server/core/objects/buffs/TimedEffect.java
package com.broll.gainea.server.core.objects.buffs;

import com.broll.gainea.server.core.GameContainer;
import com.broll.gainea.server.core.player.Player;
import com.broll.gainea.server.core.processing.GameUpdateReceiverAdapter;

public class TimedEffect extends GameUpdateReceiverAdapter {

    private GameContainer game;
    private Player owner;
    private boolean forThisTurn;
    private int rounds;

    @Override
    public void turnStarted(Player player) {
        if (owner == player || forThisTurn) {
            decrease();
        }
    }

    @Override
    public void roundStarted() {
        if (owner == null) {
            decrease();
        }
    }

    private void decrease() {
        rounds--;
        if (rounds <= 0) {
            unregister();
        }
    }

    protected void unregister() {
        this.game.getUpdateReceiver().unregister(this);
    }

    public static void forCurrentTurn(GameContainer game, TimedEffect effect) {
        effect.game = game;
        effect.forThisTurn = true;
        game.getUpdateReceiver().register(effect);
    }

    public static void forCurrentRound(GameContainer game, TimedEffect effect) {
        effect.game = game;
        game.getUpdateReceiver().register(effect);
    }

    public static void forPlayersTurn(GameContainer game, Player player, TimedEffect effect) {
        effect.game = game;
        effect.owner = player;
        game.getUpdateReceiver().register(effect);
    }

    public static void forPlayerRounds(GameContainer game, Player player, int rounds, TimedEffect effect) {
        effect.game = game;
        effect.rounds = rounds;
        effect.owner = player;
        game.getUpdateReceiver().register(effect);
    }

    public static void forGameRounds(GameContainer game, int rounds, TimedEffect effect) {
        effect.game = game;
        effect.rounds = rounds;
        game.getUpdateReceiver().register(effect);
    }

}

File: server/src/main/com/broll/gainea/server/core/objects/buffs/TimedEffect.kt
package com.broll.gainea.server.core.objects.buffs

import com.broll.gainea.server.core.Game
import com.broll.gainea.server.core.player.Player
import com.broll.gainea.server.core.processing.GameUpdateReceiverAdapter

open class TimedEffect : GameUpdateReceiverAdapter() {
    private lateinit var game: Game
    private var owner: Player? = null
    private var forThisTurn = false
    private var rounds = 0
    override fun turnStarted(player: Player) {
        if (owner === player || forThisTurn) {
            decrease()
        }
    }

    override fun roundStarted() {
        if (owner == null) {
            decrease()
        }
    }

    private fun decrease() {
        rounds--
        if (rounds <= 0) {
            unregister()
        }
    }

    protected open fun unregister() {
        game.updateReceiver.unregister(this)
    }

    companion object {
        fun forCurrentTurn(game: Game, effect: TimedEffect) {
            effect.game = game
            effect.forThisTurn = true
            game.updateReceiver.register(effect)
        }

        fun forCurrentRound(game: Game, effect: TimedEffect) {
            effect.game = game
            game.updateReceiver.register(effect)
        }

        fun forPlayersTurn(game: Game, player: Player, effect: TimedEffect) {
            effect.game = game
            effect.owner = player
            game.updateReceiver.register(effect)
        }

        fun forPlayerRounds(game: Game, player: Player, rounds: Int, effect: TimedEffect) {
            effect.game = game
            effect.rounds = rounds
            effect.owner = player
            game.updateReceiver.register(effect)
        }

        fun forGameRounds(game: Game, rounds: Int, effect: TimedEffect) {
            effect.game = game
            effect.rounds = rounds
            game.updateReceiver.register(effect)
        }
    }
}
